### ISSUE TITLE
Phase 51.1: ModalityEncoder — Specialized Multi-Modal Encoding with Shared Latent Space

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/asi-build#986]
### Overview
Implement specialized encoders for text, image, audio, and video modalities that project into a shared d-dimensional latent space. This component provides the foundation for all cross-modal operations by ensuring consistent, high-quality representations across modalities.

### Architecture

```
┌──────────────────────────────────────────────────────────────────┐
│                     ModalityEncoder                              │
│  ┌────────────┐ ┌────────────┐ ┌────────────┐ ┌──────────────┐ │
│  │ TextEncoder │ │ImageEncoder│ │AudioEncoder│ │ VideoEncoder  │ │
│  │  (BERT/    │ │  (ViT/     │ │  (Whisper/ │ │  (TimeSformer│ │
│  │   GPT-2)   │ │   CLIP-ViT)│ │   HuBERT)  │ │   /ViViT)    │ │
│  └─────┬──────┘ └─────┬──────┘ └─────┬──────┘ └──────┬───────┘ │
│        │              │              │               │          │
│  ┌─────▼──────┐ ┌─────▼──────┐ ┌─────▼──────┐ ┌─────▼───────┐ │
│  │ Projection │ │ Projection │ │ Projection │ │ Projection  │ │
│  │   Head     │ │   Head     │ │   Head     │ │   Head      │ │
│  └─────┬──────┘ └─────┬──────┘ └─────┬──────┘ └─────┬───────┘ │
│        └──────────┬───┘──────────┬───┘───────────┬───┘          │
│              ┌────▼────────────▼────────────────▼─────┐         │
│              │     Shared Latent Space (d-dim)        │         │
│              └────────────────────────────────────────┘         │
└──────────────────────────────────────────────────────────────────┘
```

### Key Components

1. **TextEncoder**: Transformer-based text encoder (BERT, GPT-2, or custom) producing token-level and sentence-level embeddings
2. **ImageEncoder**: Vision Transformer (ViT) or CNN backbone producing patch-level and image-level embeddings
3. **AudioEncoder**: Audio spectrogram transformer (AST), Whisper encoder, or HuBERT for audio representations
4. **VideoEncoder**: Temporal transformer (TimeSformer, ViViT) for spatiotemporal video representations
5. **ProjectionHead**: Learnable linear/MLP projection mapping each encoder's output to shared d-dimensional space
6. **ModalityRegistry**: Dynamic registry for adding new modality encoders at runtime
7. **BackboneFreezer**: Utilities for freezing pre-trained backbones and training only adapters/projections

### Interface

```typescript
interface IModalityEncoder {
  encode(input: ModalityInput, modality: ModalityType): Promise<LatentEmbedding>;
  encodeText(text: string | string[]): Promise<TextEmbedding>;
  encodeImage(image: ImageTensor): Promise<ImageEmbedding>;
  encodeAudio(audio: AudioTensor, sampleRate: number): Promise<AudioEmbedding>;
  encodeVideo(video: VideoTensor, fps: number): Promise<VideoEmbedding>;
  registerModality(modality: ModalityType, encoder: IEncoder): void;
  getLatentDimension(): number;
  freezeBackbone(modality: ModalityType): void;
  unfreezeBackbone(modality: ModalityType): void;
}

interface LatentEmbedding {
  tokens: Float32Array[];     // per-token embeddings
  pooled: Float32Array;       // pooled single-vector embedding
  modality: ModalityType;
  metadata: EncodingMetadata;
}

type ModalityType = "text" | "image" | "audio" | "video" | string;
```

### Implementation Details

- **Pre-trained Backbone Loading**: Support loading weights from HuggingFace-format checkpoints
- **Adapter Layers**: LoRA and prefix-tuning adapters for parameter-efficient fine-tuning
- **Mixed Precision**: FP16/BF16 encoding with automatic precision management
- **Batched Encoding**: Efficient batch processing with dynamic padding per modality
- **Embedding Normalization**: L2 normalization option for contrastive learning compatibility
- **Caching**: LRU cache for frequently encoded inputs

### Acceptance Criteria

- [ ] Text encoder produces 768-dim embeddings from input strings
- [ ] Image encoder processes 224x224+ images to patch and pooled embeddings
- [ ] Audio encoder handles variable-length audio at 16kHz
- [ ] Video encoder processes clips with temporal attention
- [ ] All encoders project to same d-dimensional shared space (configurable, default 512)
- [ ] Backbone freezing/unfreezing works correctly
- [ ] ModalityRegistry supports runtime modality addition
- [ ] Unit tests cover all encoder types with ≥90% coverage
- [ ] Benchmark: encoding throughput ≥100 samples/sec per modality on GPU

### Dependencies
- Phase 46 (Self-Supervised Learning) — contrastive pre-training foundations
- Phase 27 (Transfer Learning) — pre-trained backbone loading

### References
- Radford et al. (2021) — CLIP vision/text encoders
- Dosovitskiy et al. (2021) — ViT architecture
- Devlin et al. (2019) — BERT text encoder
- Radford et al. (2023) — Whisper audio encoder
- Arnab et al. (2021) — ViViT video transformer

*Part of [Phase 51 Planning](#985)*
